### PR TITLE
Updated Manifold to extend Projection Visualizer

### PR DESCRIPTION
**Pull request by naresh-bachwani**
This PR updates `Manifold` to extend Projection Visualizer as a part of issue #874. 

Following changes are made to Manifold:
1. It now handles 3d plot.
2. Removed duplicate code.
3. Added new parameters like `projection`

### Sample plot
```python
from yellowbrick.datasets import load_credit
from yellowbrick.features import Manifold

oz = Manifold(manifold='tsne')
oz.fit_transform(*load_credit())
oz.poof(outpath="tsne_credit.png")
```
tsne_credit

### TODOs and questions

Still to do:

- [x] Add tests.

- [x] _Is the commit message formatted correctly?_
- [ ] _Have you noted the new functionality/bugfix in the release notes of the next release?_
- [ ] _Included a sample plot to visually illustrate your changes?_
- [x] _Do all of your functions and methods have docstrings?_
- [x] _Have you added/updated unit tests where appropriate?_
- [x] _Have you updated the baseline images if necessary?_
- [x] _Have you run the unit tests using `pytest`?_
- [x] _Is your code style correct (are you using PEP8, pyflakes)?_
- [ ] _Have you documented your new feature/functionality in the docs?_

- [ ] _Have you built the docs using `make html`?_

### CODE REVIEW

File: tests/test_features/test_manifold.py [hunk 1]
```diff
@@ -117,46 +119,80 @@ def test_manifold_fit(self, mock_fit_transform):
         manifold = Manifold(target="auto")
 
         assert manifold.fit(X, y) is manifold, "fit did not return self"
-        mock_fit_transform.assert_called_once()
```
Comment (bbengfort) [inline tests/test_features/test_manifold.py]: If this is being removed, is there any reason to mock the `fit_transform` method?

File: tests/test_features/test_manifold.py [hunk 2]
```diff
@@ -166,49 +202,70 @@ def test_manifold_regression(self):
...
+        assert not hasattr(oz, "range_")
+
+        oz.fit_transform(X, y)
+        oz.finalize()
+        oz.cbar.set_ticks([])
+        assert not hasattr(oz, "classes_")
+        assert hasattr(oz, "range_")
+        self.assert_images_similar(oz, tol=15)
+
     def test_manifold_single(self):
```
Comment (bbengfort) [inline tests/test_features/test_manifold.py]: Could we also add a `test_manifold_single_3d`?

File: yellowbrick/features/manifold.py [hunk 3]
```diff
@@ -56,18 +52,13 @@
...
-
 
 ##########################################################################
 ## Manifold Embeddings
 ##########################################################################
 
-class Manifold(FeatureVisualizer):
+
+class Manifold(ProjectionVisualizer):
     """
```
Comment (bbengfort) [inline yellowbrick/features/manifold.py]: Please review the docstring updates I made to `DataVisualizer`, would you mind applying those docstring changes to `ProjectionVisualizer` and now to `Manifold` as well? (e.g. primarily the changes to the parameter descriptions).

File: yellowbrick/features/manifold.py [hunk 4]
```diff
@@ -120,30 +111,45 @@ class Manifold(FeatureVisualizer):
...
+    alpha : float, default: 0.75
         Specify a transparency where 1 is completely opaque and 0 is completely
         transparent. This property makes densely clustered points more visible.
 
     random_state : int or RandomState, default: None
         Fixes the random state for stochastic manifold algorithms.
+    
+    colorbar : bool, default: True
+        If the target_type is "continous" draw a colorbar to the right of the
+        scatter plot. The colobar axes is accessible using the cax property.
```
Comment (bbengfort) [inline yellowbrick/features/manifold.py]: Could we also please add `features_` to the `Attributes` section of the docstring?

File: yellowbrick/features/manifold.py [hunk 5]
```diff
@@ -203,24 +209,36 @@ def __init__(
...
         n_neighbors=None,
+        features=None,
+        classes=None,
         colors=None,
-        target=AUTO,
-        alpha=0.7,
+        colormap=None,
+        target_type="auto",
+        projection=2,
+        alpha=0.75,
```
Comment (bbengfort) [inline yellowbrick/features/manifold.py]: Out of curiosity, why the change from 0.7 to 0.75, to match the default in the super class?

File: yellowbrick/features/manifold.py [hunk 6]
```diff
@@ -239,21 +257,27 @@ def manifold(self, transformer):
...
             if transformer not in self.ALGORITHMS:
-                    raise YellowbrickValueError(
-                        "could not create manifold for '{}'".format(str(transformer))
-                    )
+                raise YellowbrickValueError(
+                    "could not create manifold for '{}'".format(str(transformer))
+                )
 
             # 2 components is required for 2D plots
-            # TODO: allow 3 components for 3D plots in future
```
Comment (bbengfort) [inline yellowbrick/features/manifold.py]: So excited about this!

File: yellowbrick/features/manifold.py [hunk 7]
```diff
@@ -299,179 +332,103 @@ def fit_transform(self, X, y=None):
...
-            A matrix or data frame with n instances and m features where m > 2.
+            A matrix or data frame with n instances and m features
 
         y : array-like of shape (n,), optional
             A vector or series with target values for each instance in X. This
             vector is used to determine the color of the points in X.
 
         Returns
         -------
-        self : Manifold
```
Comment (bbengfort) [inline yellowbrick/features/manifold.py]: Can we move this `returns` documentation up to `fit?`

File: yellowbrick/features/manifold.py [hunk 8]
```diff
@@ -299,179 +332,103 @@ def fit_transform(self, X, y=None):
...
-            else:
-                color_kwargs['colors'] = self.colors
-
-            self._colors = resolve_colors(**color_kwargs)
-
-        # Compute target range if colors are continuous
-        elif self._target_color_type == CONTINUOUS:
-            y = np.asarray(y)
-            self.range_ = (y.min(), y.max())
-
```
Comment (bbengfort) [inline yellowbrick/features/manifold.py]: Glorious code simplification!

File: yellowbrick/features/manifold.py [hunk 9]
```diff
@@ -299,179 +332,103 @@ def fit_transform(self, X, y=None):
...
+            
         Returns
         -------
         Xprime : array-like of shape (n, 2)
             Returns the 2-dimensional embedding of the instances.
+            
+        Note
+        ----
+        This method does not work with MDS, TSNE and SpectralEmbedding because 
+        it is yet to be implemented in sklearn.
```
Comment (bbengfort) [inline yellowbrick/features/manifold.py]: Good note!

File: yellowbrick/features/manifold.py [hunk 10]
```diff
@@ -299,179 +332,103 @@ def fit_transform(self, X, y=None):
...
-            raise NotFitted("could not determine target color type")
+        self.draw(Xp, y, **kwargs)
+        return Xp
 
-        # Draw the scatter plot with the associated colors and alpha
-        self._scatter = self.ax.scatter(X[:,0], X[:,1], **scatter_kwargs)
-        return self.ax
+    def draw(self, Xp, y=None):
+        # Calls draw method from super class which draws scatter plot.
+        super(Manifold, self).draw(Xp, y)
```
Comment (bbengfort) [inline yellowbrick/features/manifold.py]: This is probably not necessary but I like that you've added it to show explicitly that we're using the superclass draw method - very classy! (pun intended)

File: yellowbrick/features/manifold.py [hunk 11]
```diff
@@ -299,179 +332,103 @@ def fit_transform(self, X, y=None):
...
-            else:
-                self._target_color_type = CONTINUOUS
-        else:
-            self._target_color_type = self.target
-
-        if self._target_color_type not in {SINGLE, DISCRETE, CONTINUOUS}:
-            raise YellowbrickValueError((
-                "could not determine target color type "
-                "from target='{}' to '{}'"
-            ).format(self.target, self._target_color_type))
```
Comment (bbengfort) [inline yellowbrick/features/manifold.py]: Glorious code simplification!

File: yellowbrick/features/manifold.py [hunk 12]
```diff
@@ -514,49 +471,75 @@ def manifold_embedding(
         ``"tsne"``     `t-SNE <http://scikit-learn.org/stable/modules/manifold.html#t-distributed-stochastic-neighbor-embedding-t-sne>`_
         ============== ==========================
 
-    n_neighbors : int, default: 10
+    n_neighbors : int, default: None
```
Comment (bbengfort) [inline yellowbrick/features/manifold.py]: Thank you for making sure the quick method is consistent! With the docstring changes from above would you mind also ensuring that they're made down here as well!

File: tests/test_features/test_manifold.py (hunk 1 above)
Comment (naresh-bachwani) [reply tests/test_features/test_manifold.py]: My bad!

File: tests/test_features/test_manifold.py (hunk 2 above)
Comment (naresh-bachwani) [reply tests/test_features/test_manifold.py]: Sure!

File: yellowbrick/features/manifold.py (hunk 3 above)
Comment (naresh-bachwani) [reply yellowbrick/features/manifold.py]: I will have a look at it and make the necessary changes!

File: yellowbrick/features/manifold.py (hunk 4 above)
Comment (naresh-bachwani) [reply yellowbrick/features/manifold.py]: Sure!

File: yellowbrick/features/manifold.py (hunk 5 above)
Comment (naresh-bachwani) [reply yellowbrick/features/manifold.py]: Yes! This was done to match the super class.

File: yellowbrick/features/manifold.py (hunk 10 above)
Comment (naresh-bachwani) [reply yellowbrick/features/manifold.py]: Hehe! Thanks!

File: yellowbrick/features/manifold.py (hunk 12 above)
Comment (naresh-bachwani) [reply yellowbrick/features/manifold.py]: Sure!

File: yellowbrick/features/manifold.py (hunk 7 above)
Comment (naresh-bachwani) [reply yellowbrick/features/manifold.py]: My bad! Will correct it!

File: yellowbrick/features/manifold.py [hunk 13]
```diff
@@ -299,179 +361,103 @@ def fit_transform(self, X, y=None):
...
             Returns the 2-dimensional embedding of the instances.
+            
+        Note
+        ----
+        This method does not work with MDS, TSNE and SpectralEmbedding because 
+        it is yet to be implemented in sklearn.
         """
         try:
-            return self.manifold.transform(X)
+            Xp = self.manifold.transform(X)
```
Comment (bbengfort) [inline yellowbrick/features/manifold.py]: Sorry I didn't catch this before - but shouldn't this return `Xp`?

Comment (naresh-bachwani) [reply yellowbrick/features/manifold.py]: `transform`  is supposed to be the entry point of the `draw` method. And it returns Xp after the `draw` call!

Comment (naresh-bachwani) [reply yellowbrick/features/manifold.py]: But this raises yet another error! When `transform` is called followed by `fit`, `draw` is called two times. So, to counter this we need to remove `draw` call from `transform`. Hence it will be a tradeoff between these two.
To keep these both we can add conditionals based on the type of Manifold.
What do you suggest? Any other methods that can make this work?

File: yellowbrick/features/manifold.py [hunk 14]
```diff
@@ -299,179 +364,103 @@ def fit_transform(self, X, y=None):
...
-                raise YellowbrickValueError("y is required for continuous target")
-
-            # TODO manually make colorbar so we can draw it in finalize
-            scatter_kwargs["c"] = y
-            scatter_kwargs["cmap"] = self.colors or palettes.DEFAULT_SEQUENCE
-
-        else:
-            # Technically this should never be raised
-            raise NotFitted("could not determine target color type")
+        self.draw(Xp, y, **kwargs)
```
Comment (bbengfort) [inline yellowbrick/features/manifold.py]: Sorry, I couldn't see this thanks to the big diff between things!

File: yellowbrick/features/manifold.py (hunk 6 above)
Comment (rebeccabilbro) [reply yellowbrick/features/manifold.py]: Nice work guys!
